mbuild/compound.py

```diff
@@ -225,8 +247,171 @@ def particles_by_name(self, name):
+                return max(successor.rigid_id for successor in self.successors() if successor.rigid_id is not None)
+            else:
+                return self.rigid_id
+        else:
+            return None
+
+    def rigid_ids(self):
+        """Generate rigid_ids for all Particles in the Compound.
+
+        Yields
+        ------
+        int
+            The rigid_id of the next Particle in the Compound
+
+        """
+        for particle in self.particles():
+            yield particle.rigid_id
+
+    def rigid_particles(self, rigid_id=None):
+        """Generate all particles in rigid bodies.
+
+        Parameters
+        ----------
+        rigid_id : int, optional
+            Include only particles with this rigid body ID
+
+        """
+        for particle in self.particles():
+            if particle.rigid_id is not None:
+                yield particle
+
+    def set_rigid(self, name=False, rigid_id=None):
+        """Treat the Compound and all successors as a single rigid body.
+
+        Provides a unique rigid body ID to a Compound and its successors.
+        If 'name' is specified, only sucessors with this name are provided
+        with a rigid_id.  This is useful for Compounds where only some of
+        the particles should be rigid (such as a tethered nanoparticle with
+        a rigid core).
+
+        Parameters
+        ----------
+        name : str, optional
+            Include only particles with this name in the rigid body.
+        rigid_id : int, optional
+            Manually set the rigid body ID for this body.
+            Note: This is not recommended, as mBuild keeps track of all
+            rigid body ID's and manually overriding this can lead to 
+            unwanted behavior.
+
+        """
+        if rigid_id is None:
+            if self.root.max_rigid() is not None:
+                rigid_id = self.root.max_rigid() + 1
+            else:
+                rigid_id = 0
+        self.rigid_id = rigid_id
+        for successor in self.successors():
+            if name and successor.name != name:
+                pass
+            else:
+                successor.rigid_id = rigid_id
+        if rigid_id and rigid_id > self.root.max_rigid() + 1:
+            warn("Specified 'rigid_id' > max rigid_id + 1. This will lead to inconsistent rigid body numbering. Consider running reset_rigid().")
+
+    def _set_rigid_only(self, rigid_id):
+        """Resets the rigid_id of all rigid Particles in a Compound
+
+        Treats the Compound and all successors that already have a rigid_id as
+        a single rigid body.
+        """
+        self.rigid_id = rigid_id
+        for successor in self.successors():
+            if successor.rigid is not False:
+                if successor.rigid_id is not None:
+                    successor.rigid_id += rigid_id
+                else:
+                    successor.rigid_id = rigid_id
+
+    def create_rigid_bodies(self, name, particle_name=None):
```

Comment:
I think we had talked about merging this with `set_rigid`, right? Did you think of a use case where this won't work?